### PR TITLE
create: add TCMU Ring Buffer size settable

### PR DESCRIPTION
**Pull request by lxbsz**
$ gluster-block create dht0/block33 rb_size 60 ha 3 192.168.195.137,192.168.195.134,192.168.195.135 1G --json-pretty
{
  "IQN":"iqn.2016-12.org.gluster-block:1e9d26a4-28d7-43a5-a2dc-b7d08a6cbedd",
  "PORTAL(S)":[
    "192.168.195.137:3260",
    "192.168.195.134:3260",
    "192.168.195.135:3260"
  ],
  "RESULT":"SUCCESS"
}

$ cat block-meta/block32
VOLUME: dht0
GBID: da787d30-ac8a-420c-8907-09b494e07ebf
HA: 3
ENTRYCREATE: INPROGRESS
SIZE: 1073741824
RB_SIZE: 60MB
ENTRYCREATE: SUCCESS
192.168.195.137: CONFIGINPROGRESS
192.168.195.134: CONFIGINPROGRESS
192.168.195.135: CONFIGINPROGRESS
192.168.195.134: CONFIGSUCCESS
192.168.195.137: CONFIGSUCCESS
192.168.195.135: CONFIGSUCCESS

$ targetcli /backstores/user:glfs/block32/ get attribute max_data_area_mb
max_data_area_mb=60 [ro]

Signed-off-by: Xiubo Li <xiubli@redhat.com>

### CODE REVIEW

**Comment by pkalever** [comment]
@lxbsz Can you please add new capability for this ?

**Comment by pkalever** [comment]
@pranithk how do we make sure the new member in the rpc blockCreate structure wont break the compatibility ? Simply increasing rpc version number ? Can we have your insight here ?

**Comment by pkalever** [comment]
@pranithk @vbellur this patch introduces a new option **'rb_size'** to gluster-block create command.
IMO we will have to make some betterments to option naming, may be something like rbsize, rb-size, buffsize, iobuffsize or with some '-' in between them ?

Please suggest.

**Comment by pranithk** [comment]
@pkalever for version compatibility of the structure on the network, there should be one more rpc structure in .x, based on the capability it should choose the new rpc vs old rpc call.
Here are the cases:
If we get a create request with rb-size but the version compatibility fails in such a way that the HA count can not be met, then fail the operation

If we get a create request with no rb-size and the version compatibility says there are some nodes without rb-size change, then choose the old rpc which uses blockCreate() rpc.

If we get a create request and the version compatibility says that all the rpcs have rb-size change, then choose the new rpc which uses blockCreatev2() rpc which has rb-size.

**Comment by pranithk** [comment]
@pkalever I personally like ring-buffer-size

**Comment by pkalever** [comment]
@pranithk Is this how we handle in glusterd ? everytime we have a change in the rpc structure, have a new copy of structure ?

I personally never got a chance to explored the potential of program number, version number and  procedure number in this kind of scenarios. 

At-least from theory I was under the impression that we should increment GLUSTER_BLOCK_VERS and  GLUSTER_BLOCK_CLI_VERS ?

**Comment by pkalever** [comment]
@amarts is that something that you can help with :-)

File: rpc/block_svc_routines.c [hunk 1]
```diff
@@ -3029,11 +3035,14 @@ block_create_cli_1_svc_st(blockCreateCli *blk, struct svc_req *rqstp)
   }
 
   GB_METAUPDATE_OR_GOTO(lock, glfs, blk->block_name, blk->volume,
-                        errCode, errMsg, exist, "SIZE: %zu\nENTRYCREATE: SUCCESS\n", blk->size);
+                        errCode, errMsg, exist,
+                        "SIZE: %zu\nRB_SIZE: %s\nENTRYCREATE: SUCCESS\n",
```
Comment (pkalever) [inline rpc/block_svc_routines.c]: @pranithk We are feeding RB_SIZE to metafile, should we show this as part of in the gluster-block info ?

**Comment by pkalever** [comment]
@pranithk looks like we have missed this structure change precautions earlier with below commits too:

commit 7008073e1d00c1af0dc5913cdf5d94372f10088d
Author: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>
Date:   Wed Nov 29 17:44:50 2017 +0530

    delete: support [unlink-storage <yes|no>] option
    
    $ gluster-block help
    gluster-block (0.3)
    usage:
      gluster-block <command> <volname[/blockname]> [<args>] [--json*]
    
    commands:
    [...]
    
      delete  <volname/blockname> [unlink-storage <yes|no>] [force]
            delete block device.
    [...]
    supported JSON formats:
      --json|--json-plain|--json-spaced|--json-pretty
    
    Change-Id: I64bc99a8519be6f90a7e8bc5558b0d7518661995
    Fixes: #19
    Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

commit 450f7e89116ed104f092dd1a709c0f057150a4f7 (refs/heads/type)
Author: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>
Date:   Wed Feb 7 21:40:13 2018 +0530

    create: support [storage <filename>] option
    
    $ gluster-block help
    gluster-block (0.3)
    usage:
      gluster-block <command> <volname[/blockname]> [<args>] [--json*]
    
    commands:
      create  <volname/blockname> [ha <count>]
                                  [auth <enable|disable>]
                                  [prealloc <full|no>]
                                  [storage <filename>]
                                  <host1[,host2,...]> <size>
            create block device [defaults: ha 1, auth disable, prealloc no, size in bytes]
    [...]
    supported JSON formats:
      --json|--json-plain|--json-spaced|--json-pretty
    
    Fixes: #18
    Change-Id: I5d9b4abb58596bd2297cffb294c6d31f752d53e9
    Signed-off-by: Prasanna Kumar Kalever <prasanna.kalever@redhat.com>

Look like we need to address/fix this as well ?

File: rpc/block_svc_routines.c (hunk 1 above)
Comment (pranithk) [reply rpc/block_svc_routines.c]: @pkalever Will it be relevant when dynamic ring buffer size patch comes? If yes, then we should, otherwise no IMHO

**Comment by pranithk** [comment]
> @pranithk Is this how we handle in glusterd ? everytime we have a change in the rpc structure, have a new copy of structure ?

For glusterd/cli, it is fine as we expect it to have same version. For the fops between glusterd/glusterd, we use dictionary to add extra variables. So one thing to do for future proofing is to add the opaque xdata<> kind of argument even for all the requests next time when we have to change. So if we decide to go ahead with this feature, we should do that for Create op.

**Comment by pkalever** [comment]
@lxbsz @pranithk have updated the patch please review

**Comment by pkalever** [comment]
@lxbsz Apologies for closing this, it was unfortunate. Later tried to reopen but I got permission denied.

I have pushed a new PR here:
https://github.com/gluster/gluster-block/pull/64

**Comment by lxbsz** [comment]
@pkalever Don't worry. Thanks for your great work to push this.

**Pull request closed without merging**